File: 0057-insert-interval/0057-insert-interval.go
func insert(intervals [][]int, newInterval []int) [][]int {
    if len(intervals) == 0 {
        return append([][]int{}, newInterval)
    }
    
    if intervals[0][1] < newInterval[0] {
        return append(append([][]int{},intervals[0]),
                      insert(intervals[1:], newInterval)...)
    }
    
    if intervals[0][1] >= newInterval[0] && intervals[0][1] <= newInterval[1] {
        return append([][]int{},
                      insert(intervals[1:],
                             []int{min(intervals[0][0], newInterval[0]),
                                   newInterval[1]})...)
    }
    
    if intervals[0][1] >= newInterval[1] && intervals[0][0] <= newInterval[1] {
        intervals[0][0] = min(intervals[0][0], newInterval[0])
        return intervals
    }
    
    if intervals[0][0] > newInterval[1] {
        return append(append([][]int{}, newInterval), intervals...)
    }
    
    return nil
    
}

func min(a int, b int) int {
    if a < b {
        return a
    }
    return b
}

File: 0057-insert-interval/0057-insert-interval.rb
# @param {Integer[][]} intervals
# @param {Integer[]} new_interval
# @return {Integer[][]}
def insert(intervals, new_interval)
  new_left, new_right = new_interval

  # First, binary search for intervals we know for sure is beyond the new one
  right_index = intervals.bsearch_index do |cur_left, cur_right|
    cur_left > new_right
  end
  # Assume not found means that there are no intervals beyond new_interval.
  # This also helps cover the base case of intervals = []
  right_index = intervals.length if right_index.nil?
  
  # Will append this to the result later
  right_intervals = intervals[right_index..]
  
  # Now check for merges along left
  insert_index = right_index
  (right_index - 1).downto(0) do |index|
    cur_left, cur_right = intervals[index]
    if cur_right < new_left
      # Not in range, so we're done
      break
    end
    
    # Update the new interval
    insert_index = index
    new_right = [cur_right, new_right].max
    new_left = [cur_left, new_left].min
  end
  
  # Now compile the result by inserting the new interval
  # between left and right intervals that weren't merged.
  result = intervals[0...insert_index]
  result.push([new_left, new_right])
  result + right_intervals
end